source/crud/bulk-write.md

```diff
@@ -0,0 +1,810 @@
+    "code": Optional<Int32>,
+    "errmsg": Optional<String>,
+    "errInfo": Optional<Document>,
+    "n": <Int32>,
+    "nModified": Optional<Int32>,
+    "upserted": Optional<Document with "_id" field>
+}
+```
+
+If an error occurred (i.e. the value for `ok` is 0), the `code`, `errmsg`, and optionally
+`errInfo` fields will be populated with details about the failure.
+
+If the write succeeded, (i.e. the value for `ok` is 1), `n`, `nModified`, and `upsertedId` will be
+populated with the following values based on the type of write:
+
+| Response Field | Insert | Update | Delete |
+| -------------- | ------ | ------ | ------ |
+| `n` | The number of documents that were inserted. | The number of documents that matched the filter. | The number of documents that were deleted. |
+| `nModified` | Not present. | The number of documents that were modified. | Not present. |
+| `upserted` | Not present. | A document containing the `_id` value for the upserted document. Only present if an upsert took place. | Not present. |
+
+Note that the responses do not contain information about the type of operation that was performed.
+Drivers may need to maintain the user's list of write models to infer which type of result should
+be recorded based on the value of `idx`.
+
+### Handling Insert Results
+
+Unlike the other result types, `InsertOneResult` contains an `insertedId` field that is generated
+driver-side, either by recording the `_id` field present in the user's insert document or creating
+and adding one. Drivers MUST only record these `insertedId`s in a `BulkWriteResult` when a
+successful response for the insert operation (i.e. `{ "ok": 1, "n": 1 }`) is received in the
+results cursor. This ensures that drivers only report an `insertedId` when it is confirmed that the
+insert succeeded.
+
+## Handling Errors
+
+### Top-Level Errors
+
+A top-level error is any error that occurs that is not the result of a single write operation
+failing or a write concern error. Examples include network errors that occur when communicating
+with the server, command errors (`{ "ok": 0 }`) returned from the server, client-side errors, and
+errors that occur when attempting to perform a `getMore` to retrieve results from the server.
+
+When a top-level error is encountered and individual results and/or errors have already been
+observed, drivers MUST embed the top-level error within a `BulkWriteException` as the `error` field
+to retain this information. Otherwise, drivers MAY throw an exception containing only the top-level
+error.
+
+Encountering a top-level error MUST halt execution of a bulk write for both ordered and unordered
+bulk writes. This means that drivers MUST NOT attempt to retrieve more responses from the cursor or
+execute any further `bulkWrite` batches and MUST immediately throw an exception. If the results
+cursor has not been exhausted on the server when a top-level error occurs, drivers MUST send the
+`killCursors` command to attempt to close it.
```

Comment:
```suggestion
`killCursors` command to attempt to close it. The result from the `killCursors` command MAY be ignored.
```

I expect sending `killCursors` is a best-effort way to clean-up the cursor. If it fails, ignore the error. Suggested sentence is copied from [Find, getMore and killCursors commands](https://github.com/mongodb/specifications/blob/611b12ccbdd012dcd9ab2877a32200b3835c97af/source/find_getmore_killcursors_commands.rst#L446).

Reply:
updated